src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioSymbolNavigationService.cs

```diff
@@ -121,16 +124,15 @@ public bool TryNavigateToSymbol(ISymbol symbol, Project project, OptionSet optio
             // Check whether decompilation is supported for the project. We currently only support this for C# projects.
             if (project.LanguageServices.GetService<IDecompiledSourceService>() != null)
             {
+                var eulaService = project.Solution.Workspace.Services.GetRequiredService<IDecompilerEulaService>();
                 allowDecompilation = project.Solution.Workspace.Options.GetOption(FeatureOnOffOptions.NavigateToDecompiledSources) && !symbol.IsFromSource();
-                if (allowDecompilation && !project.Solution.Workspace.Options.GetOption(FeatureOnOffOptions.AcceptedDecompilerDisclaimer))
+                if (allowDecompilation && !ThreadingContext.JoinableTaskFactory.Run(() => eulaService.IsAcceptedAsync(cancellationToken)))
                 {
-                    var notificationService = project.Solution.Workspace.Services.GetService<INotificationService>();
+                    var notificationService = project.Solution.Workspace.Services.GetRequiredService<INotificationService>();
                     allowDecompilation = notificationService.ConfirmMessageBox(ServicesVSResources.Decompiler_Legal_Notice_Message, ServicesVSResources.Decompiler_Legal_Notice_Title, NotificationSeverity.Warning);
                     if (allowDecompilation)
                     {
-                        var workspace = project.Solution.Workspace;
-                        workspace.TryApplyChanges(workspace.CurrentSolution.WithOptions(workspace.Options
-                            .WithChangedOption(FeatureOnOffOptions.AcceptedDecompilerDisclaimer, true)));
+                        ThreadingContext.JoinableTaskFactory.Run(() => eulaService.MarkAcceptedAsync(cancellationToken));
                     }
                 }
```

Comment:
if decompilatin is not allowed, what happens here

Reply:
What do you mean by "not allowed" in this context?

Reply:
confirmation box shows.  user clicks 'no'.  so we don't persist the eula acceptance.  won't they get into a prompt-loop teh next time they do anything?

Reply:
If decompilation is not allowed, meaning the user clicked 'No', then nothing happens. Put another way, this pull request does not change #42638. It only improves the experience when users hit 'Yes', which is presumably the case because the dialog only shows after a user manually enables this feature.

Reply:
> which is presumably the case because the dialog only shows after a user manually enables this feature.

but if a user manually enabled the feature... then they should have been prompted for the eula *then*.  and only if they accepted the eula should we even store the value that they want decompilation.  if they did not accept the eual then, then we shouldhave never set the 'true' bit for decompilation.  

I honestly don't get this.  it doesn't make sense to me (either in terms of user experience or in terms of the code).

Reply:
basically, i do not understand how the setting to decompile could ever be true, while the setting for the eula is ever false.  not at this location in the code.  We're in symbol-nav.  These settings should be in a consistent state long before we ever get here.

Reply:
The EULA is not presented when the option is checked in options, because that could have occurred while the user had a different solution open. The option is checked the first time navigation to a symbol in metadata occurs, where it can be checked in the context of persistence for the current solution.

Reply:
@mikadumont @jinujoseph  on this.  i'm just not getting it.  there' no reason i would expect or want to tie this option to a solution.  

> because that could have occurred while the user had a different solution open. 

If i go to option and turn on decompilaiton .i should get a prompt.  that prompt should be remembered forever.  Best case is that it's associated with my account and roamed.  Worst case it's remembered on the local machine.  But i can't understand any sort of design that ties this to somethinglike the solution or to navigation.  That makes no sense on any level, and does not give the user any sort of expected experience around decompilation that is sensible.

The fallout of that is the types of experiences i've been running into :-/

Reply:
@CyrusNajmabadi This is not our call to make on the public issue tracker. If you want to change the location where the message is presented, that request must go through an internal legal approval process.

Reply:
Talked to @mikadumont offline about this.  We're almost totally ok with where this PR brings us in terms of user expeirence.  One thing we want to change though is that if the user clicks 'no', we remember that choice of theirs and we disable "enable navigation to decompiled...".  That way they don't get pestered after informing us of their decision.

Then, if they want to enable this later, they can do so through tools|options.  

This is also a viable approach if we ever enable this option by default.  If we do, users will get the prompt on first-nav, but we will respect and persist their decision here on what they want their experience to be.

Tagging @jinujoseph as well.